python/ray/air/checkpoint.py

```diff
@@ -454,6 +468,34 @@ def from_directory(cls, path: str) -> "Checkpoint":
 
         return checkpoint
 
+    @classmethod
+    def from_directory_populator(cls, populator: Callable[[str], Any]):
+        """Create a checkpoint from a populator function.
+
+        The populator function receives one argument, a string pointing to a directory.
+        The function can proceed to write data into this directory.
+
+        This method will then return a Checkpoint created from this directory.
+
+        Args:
+            populator: Function taking a directory path as an input. The populator
+                function can write checkpoint data into this directory.
+
+        Returns:
+            Checkpoint object containing data written to the directory.
+
+        Examples:
+            >>> checkpoint = Checkpoint.from_directory_populator(  # doctest: +SKIP
+            ...     lambda dir: (Path(dir) / "data.txt").write_text("Data")
+            ... )
+
+        """
+        temp_directory = tempfile.mkdtemp()
```

Comment:
Do we want to discuss the optionality of being able to create a checkpoint in the "final location" or would having the ability to just move a `_managed` checkpoint be sufficient for performance?

Reply:
I think that is an implementation detail and a possible future optimization, if needed. For performance it will only be important that this is created on the disk where it is moved to (or a tmpfs). We may obtain this from the air session in the future if it makes sense.

For now I think we're good with a temporary directory.